app/web/src/newhotness/logic_composables/component_actions.ts

```diff
@@ -0,0 +1,265 @@
+    if (!actionViewList.data.value?.actions?.length || !component.value?.id) {
+      return {};
+    }
+
+    const result: Record<ActionPrototypeId, ActionProposedView> = {};
+    for (const action of actionViewList.data.value.actions) {
+      if (action.componentId === component.value.id) {
+        // NOTE(nick): this assumes that there can be one action for a given prototype and component.
+        // As of the time of writing, this is true, but multiple actions per prototype and component
+        // aren't disallowed from the underlying graph's perspective. Theorhetically, you could
+        // enqueue two refreshes back-to-back. What then? I don't think we'll expose an interface to
+        // do that for awhile, so this should be sufficient.
+        result[action.prototypeId] = action;
+      }
+    }
+    return result;
+  });
+
+  const actionIsRunning = (actionId?: ActionId | null) => {
+    if (!actionId) return false;
+    const state = actionViewList.data.value?.actions.find(
+      (action) => action.id === actionId,
+    )?.state;
+    if (!state) return false;
+    return [ActionState.Dispatched, ActionState.Running].includes(state);
+  };
+
+  // Helpers for Refresh Actions specifically
+  const refreshActionPrototype = computed(() => {
+    return (
+      actionPrototypeViews.value.find(
+        (action: ActionPrototypeView) => action.kind === ActionKind.Refresh,
+      ) ?? null
+    );
+  });
+
+  const refreshAction = computed(() => {
+    if (!refreshActionPrototype.value?.id) return null;
+    return actionByPrototype.value[refreshActionPrototype.value.id] ?? null;
+  });
+
+  const refreshActionRunning = computed(() => {
+    return refreshAction.value
+      ? actionIsRunning(refreshAction.value?.id)
+      : false;
+  });
+
+  const refreshEnabled = computed(
+    () =>
+      !!(
+        refreshActionPrototype.value &&
+        component.value?.hasResource &&
+        componentExistsOnHead.value
+      ),
+  );
+
+  const runRefreshHandler = () => {
+    const directRefreshApi = useApi();
+
+    const executeRefresh = async () => {
+      if (!component.value?.id || refreshActionRunning.value) return;
+      const call = directRefreshApi.endpoint(routes.RefreshAction, {
+        componentId: component.value.id,
+      });
+      directRefreshApi.setWatchFn(() => refreshAction.value?.state);
+      await call.put({});
+    };
+
+    const bifrosting = computed(() => directRefreshApi.bifrosting.value);
+
+    return {
+      executeRefresh,
+      bifrosting,
+    };
+  };
+
+  const toggleActionHandler = (
+    actionPrototypeView: ActionPrototypeView,
+    actionId?: MaybeRefOrGetter<ActionId | undefined>,
+  ) => {
+    const removeApi = useApi();
+    const addApi = useApi();
+    const refreshApi = useApi();
+    // Track which API is currently active as we could call 3 different APIs and need to propagate
+    // the bifrosting status reactively for the correct one
+    const activeApi = ref(addApi);
```

Comment:
This is clever and makes sense to me. Something feels clunky, but I can't quite put my finger on it. It's certainly not how you used it. In fact, this is probably the most ideal way to have dealt with multiple bifrosting values. I'll think on this for the future, but looks great to me.